### PR TITLE
AB#139145: Solvent surfer

### PR DESCRIPTION
**Pull request by JoeHeeley2**
# Overview

Adding solvent surfer functionality. Code is mostly found in the sources/blueprints/solvent_PCA directory. There are some additions to other scripts too.

### CODE REVIEW

File: Webapp/sources/blueprints/save_reaction/routes.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
 import uuid
 from datetime import datetime
 
-import magic
+#import magic
```
Comment (JoeDavies-6) [inline Webapp/sources/blueprints/save_reaction/routes.py:11]: restore magic

File: Webapp/sources/blueprints/solvent_PCA/data/Reaction_data_references.docx
Comment (AndyRae) [inline Webapp/sources/blueprints/solvent_PCA/data/Reaction_data_references.docx:1]: Is the .docx file necessary? I can't see it used in code, if it's docs then a markdown file?

File: Webapp/sources/templates/_saved_pca_graphs.html [hunk 2]
```diff
@@ -0,0 +1,16 @@
+<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.4.0/css/font-awesome.min.css">
```
Comment (AndyRae) [inline Webapp/sources/templates/_saved_pca_graphs.html:1]: There's already a version of this loaded in the `base.html`

**Review by AndyRae — CHANGES_REQUESTED** [review]

File: Webapp/sources/blueprints/solvent_PCA/utils.py [hunk 3]
```diff
@@ -0,0 +1,378 @@
+import ast
```
[the remark this answers is not in the document]
Comment (AndyRae) [reply Webapp/sources/blueprints/solvent_PCA/utils.py:1]: Having looked through it all - is it library code from elsewhere? Feel like I'm missing context, but I'm super hesitant to put it all into blueprints

[the remark this answers is not in the document]
Comment (JoeHeeley2) [reply Webapp/sources/blueprints/solvent_PCA/utils.py:1]: The code in solvent_PCA/interactive has all been adapted from from a separate application, the rest has been written to connect it to the front end. I'd be happy to move everything other than routes elsewhere, do you think services would be the best place for it?

File: Webapp/sources/blueprints/save_reaction/routes.py (hunk 1 above)
Comment (JoeHeeley2) [reply Webapp/sources/blueprints/save_reaction/routes.py]: restored

File: Webapp/sources/blueprints/solvent_PCA/interactive/Dataset.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        self.data = copy(self.original_data)
+        self.label_name = self.attribute_names[self.label_index]
+        self.ignored_attributes.append(self.label_name)
+        self.mask = []
+        for i, name in enumerate(self.attribute_names):
+            if name not in self.ignored_attributes:
+                self.mask.append(i)
+        self.update_data()
+        self.normalize()
+        # self.update_data()
```
[the remark this answers is not in the document]
Comment (JoeHeeley2) [reply Webapp/sources/blueprints/solvent_PCA/interactive/Dataset.py]: I've removed all lines of code that have been commented out, including in other files

File: Webapp/sources/blueprints/solvent_PCA/data/Reaction_data_references.docx
Comment (JoeHeeley2) [reply Webapp/sources/blueprints/solvent_PCA/data/Reaction_data_references.docx:1]: these have been removed

File: Webapp/sources/models/PCA_graph.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
+from sources.extensions import db
+
+from .base import Model
+
+# metadata = db.Model.metadata
+
+
+class PCAGraph(Model):
```
[the remark this answers is not in the document]
Comment (JoeHeeley2) [reply Webapp/sources/models/PCA_graph.py:8]: How do you connect to the live database to make the migrations? can you connect via ssh?

File: Webapp/sources/templates/_saved_pca_graphs.html (hunk 2 above)
Comment (JoeHeeley2) [reply Webapp/sources/templates/_saved_pca_graphs.html]: removed this from the file

File: Webapp/sources/models/PCA_graph.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (AndyRae) [reply Webapp/sources/models/PCA_graph.py:8]: There's no need to - the migrations run on app start, see the `startup.sh` file in project root.

File: Webapp/sources/blueprints/solvent_PCA/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (AndyRae) [reply Webapp/sources/blueprints/solvent_PCA/utils.py:1]: Yep that would make sense - essentially we want to make everything in blueprints are just `routes.py` (not immediately, but over time, and certainly not add further logic there). 
And then make the routes themselves "thinner", so all the business logic of the application is in services. So the blueprints are just coordinating the request, and calling services.

[the remark this answers is not in the document]
Comment (JoeHeeley2) [reply Webapp/sources/blueprints/solvent_PCA/utils.py]: all functions have now been moved into a solvent_PCA directory in services

**Comment by JoeHeeley2** [comment]
which file needs updating to account for new dependencies?

**Comment by JoeDavies-6** [comment]
> which file needs updating to account for new dependencies?

I think it's this one - https://github.com/AI4Green/AI4Green/blob/main/pyproject.toml

File: Webapp/sources/blueprints/reaction_table/reaction_classification.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
+from rdkit import Chem
```
Comment (JoeDavies-6) [inline Webapp/sources/blueprints/reaction_table/reaction_classification.py:1]: this file could probably go to services as well

File: Webapp/sources/models/PCA_graph.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    PCA plots can be replotted from:
+    r_class: reaction class selected in reaction class dropdown
+    colour_selected: colour in colour dropdown
+    control_points: defined points for constrained PCA
+
+    """
+
+    __tablename__ = "PCA"
+
+    id = db.Column(db.Integer, primary_key=True)
```
Comment (JoeDavies-6) [inline Webapp/sources/models/PCA_graph.py:19]: Do we need date and time of creation? 
Possibly, just future work for another PR but do we want to link this to a WorkBook to enable sharing of graphs?

File: Webapp/sources/blueprints/reaction_table/reaction_classification.py (hunk 6 above)
Comment (JoeHeeley2) [reply Webapp/sources/blueprints/reaction_table/reaction_classification.py]: done

File: Webapp/sources/blueprints/reaction_table/routes.py [hunk 8]
```diff
@@ -4,19 +4,20 @@
...
 from flask import current_app, jsonify, render_template, request
 from flask_login import login_required
 from rdkit import Chem
+from rdkit.Chem import Descriptors
 from sources import models, services
 from sources.auxiliary import abort_if_user_not_in_workbook, smiles_symbols
 from sources.dto import ReactionNoteSchema
 
 from . import reaction_table_bp
+from Webapp.sources.services.reaction_classification import classify_reaction
```
Comment (JoeDavies-6) [inline Webapp/sources/blueprints/reaction_table/routes.py:20]: guessing this is pycharm's refactor->move?  more consistent to use as: services.reaction_classification.classify()

File: Webapp/sources/models/PCA_graph.py (hunk 7 above)
Comment (JoeHeeley2) [reply Webapp/sources/models/PCA_graph.py:19]: yeah I included both to make it consistent with the reactions. 
Possibly, I guess we could talk to some organic chemists about specific use cases and whether they would bother collaborating using the solvent surfer or just use them personally

File: Webapp/sources/services/PCA_graph.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
+from flask_login import current_user
```
Comment (JoeDavies-6) [inline Webapp/sources/services/PCA_graph.py:1]: you may need to add all new files to services/__init__ to make them findable

File: Webapp/sources/blueprints/solvent_PCA/routes.py [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+from flask import Response, jsonify, render_template, request
+from flask_login import login_required
+from sources import models
+from sources.services.person import from_current_user_email
+from sources.services.PCA_graph import list_user_graphs, PCAgraph_from_id
+from sources.auxiliary import get_notification_number, get_workgroups
+from sources.extensions import db
+
+from . import solvent_PCA_bp
+from Webapp.sources.services.solvent_PCA.embedding import get_PCA, get_r_class_descriptors
```
Comment (JoeDavies-6) [inline Webapp/sources/blueprints/solvent_PCA/routes.py:15]: same with these imports - I've been sticking to the pattern of importing services and then just using services.domain.function()

File: Webapp/sources/services/solvent_PCA/graph_utils.py [hunk 11]
```diff
@@ -0,0 +1,447 @@
...
+from bokeh.plotting import figure
+from bokeh.transform import linear_cmap
+
+from .javascript_callbacks import (
+    get_exit_interactive_callback,
+    get_interactive_setup_callback,
+    get_reset_interactive_callback,
+    get_save_callback,
+)
+from Webapp.sources.blueprints.main.routes import get_custom_colours
```
Comment (JoeDavies-6) [inline Webapp/sources/services/solvent_PCA/graph_utils.py:26]: and here

File: Webapp/sources/models/PCA_graph.py (hunk 7 above)
Comment (JoeDavies-6) [reply Webapp/sources/models/PCA_graph.py:19]: Hadn't noticed in models/reactions. I don't think we ever use Reaction.date_reaction, we should remove that at some point, and I reckon this one too.
Sounds like a good idea for use cases.

**Comment by AndyRae** [comment]
> > which file needs updating to account for new dependencies?
> 
> I think it's this one - https://github.com/AI4Green/AI4Green/blob/main/pyproject.toml

yep, see: https://python-poetry.org/docs/cli/#add

**Review by AndyRae — APPROVED** [review]

**Comment by AndyRae** [comment]
LGTM - if possible could you wait until Monday to deploy to production? Obviously Friday afternoon is never nice, but as this is migrations and I think the 3rd one, would prefer it be Monday incase it goes wrong. Thanks!

**Comment by JoeHeeley2** [comment]
will do, hope you have a good weekend!
________________________________

**Review by JoeDavies-6 — DISMISSED** [review]

**Comment by JoeHeeley2** [comment]
Sorry, just realised I didn't push the updated .toml file

**Review by JoeDavies-6 — APPROVED** [review]

**Review by AndyRae — APPROVED** [review]

**Pull request merged**